planner/core/logical_plan_builder.go

```diff
@@ -1138,96 +1138,193 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
+		proj.SetChildren(setOprPlan)
+		schema := expression.NewSchema(setOprPlan.Schema().Clone().Columns[:oldLen]...)
 		for _, col := range schema.Columns {
 			col.UniqueID = b.ctx.GetSessionVars().AllocPlanColumnID()
 		}
-		proj.names = unionPlan.OutputNames()[:oldLen]
+		proj.names = setOprPlan.OutputNames()[:oldLen]
 		proj.SetSchema(schema)
 		return proj, nil
 	}
+	return setOprPlan, nil
+}
 
-	return unionPlan, nil
+func (b *PlanBuilder) buildIntersect(ctx context.Context, selects []*ast.SelectStmt) (LogicalPlan, *ast.SetOprType, error) {
+	leftPlan, err := b.buildSelect(ctx, selects[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(selects) == 1 {
+		return leftPlan, selects[0].AfterSetOperator, nil
+	}
+
+	columnNums := leftPlan.Schema().Len()
+	for i := 1; i < len(selects); i++ {
+		rightPlan, err := b.buildSelect(ctx, selects[i])
+		if err != nil {
+			return nil, nil, err
+		}
+		if rightPlan.Schema().Len() != columnNums {
+			return nil, nil, ErrWrongNumberOfColumnsInSelect.GenWithStackByArgs()
+		}
+		joinPlan := LogicalJoin{JoinType: SemiJoin}.Init(b.ctx, b.getSelectOffset())
+		joinPlan.SetChildren(leftPlan, rightPlan)
+		joinPlan.SetSchema(leftPlan.Schema())
+		joinPlan.names = make([]*types.FieldName, leftPlan.Schema().Len())
+		copy(joinPlan.names, leftPlan.OutputNames())
+		for j := 0; j < len(rightPlan.Schema().Columns); j++ {
+			leftCol, rightCol := leftPlan.Schema().Columns[j], rightPlan.Schema().Columns[j]
+			eqCond, err := expression.NewFunction(b.ctx, ast.NullEQ, types.NewFieldType(mysql.TypeTiny), leftCol, rightCol)
+			if err != nil {
+				return nil, nil, err
+			}
+			joinPlan.EqualConditions = append(joinPlan.EqualConditions, eqCond.(*expression.ScalarFunction))
+		}
+		leftPlan, err = b.buildDistinct(joinPlan, leftPlan.Schema().Len())
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return leftPlan, selects[0].AfterSetOperator, nil
 }
 
-// divideUnionSelectPlans resolves union's select stmts to logical plans.
-// and divide result plans into "union-distinct" and "union-all" parts.
-// divide rule ref: https://dev.mysql.com/doc/refman/5.7/en/union.html
-// "Mixed UNION types are treated such that a DISTINCT union overrides any ALL union to its left."
-func (b *PlanBuilder) divideUnionSelectPlans(ctx context.Context, selects []*ast.SelectStmt) (distinctSelects []LogicalPlan, allSelects []LogicalPlan, err error) {
-	firstUnionAllIdx, columnNums := 0, -1
-	// The last slot is reserved for appending distinct union outside this function.
-	children := make([]LogicalPlan, len(selects), len(selects)+1)
-	for i := len(selects) - 1; i >= 0; i-- {
-		stmt := selects[i]
-		if firstUnionAllIdx == 0 && stmt.IsAfterUnionDistinct {
-			firstUnionAllIdx = i + 1
+func (b *PlanBuilder) buildExcept(ctx context.Context, selects []LogicalPlan, afterSetOpts []*ast.SetOprType) (LogicalPlan, error) {
```

Comment:
Please add some comments for `buildIntersect` and `buildExcept`.

Reply:
Done.